File: Examples/ButtonsExample/UIView+Theming.m
//
//  UIView+Theming.m
//  ButtonsExample
//
//  Created by Eric Horacek on 12/27/14.
//  Copyright (c) 2014 Eric Horacek. All rights reserved.
//

@import Motif;

#import "ThemeSymbols.h"

#import "UIView+Theming.h"

NS_ASSUME_NONNULL_BEGIN

@implementation UIView (Theming)

+ (void)load {
    [self
        mtf_registerThemeProperty:ThemeProperties.borderWidth
        requiringValueOfClass:NSNumber.class
        applierBlock:^(NSNumber *width, UIView *view, NSError **error) {
            view.layer.borderWidth = width.floatValue;
            return YES;
        }];

    [self
        mtf_registerThemeProperty:ThemeProperties.borderColor
        requiringValueOfClass:UIColor.class
        applierBlock:^(UIColor *color, UIView *view, NSError **error) {
            view.layer.borderColor = color.CGColor;
            return YES;
        }];
    
    [self
        mtf_registerThemeProperty:ThemeProperties.cornerRadius
        requiringValueOfClass:NSNumber.class
        applierBlock:^(NSNumber *cornerRadius, UIView *view, NSError **error) {
            view.layer.cornerRadius = cornerRadius.floatValue;
            return YES;
        }];
    
    [self
        mtf_registerThemeProperty:ThemeProperties.backgroundColor
        requiringValueOfClass:UIColor.class
        applierBlock:^(UIColor *color, UIView *view, NSError **error) {
            view.backgroundColor = color;
            return YES;
        }];
}

@end

NS_ASSUME_NONNULL_END

File: Examples/SwiftButtonsExample/UIView+Theming.swift
//
//  UIView+Theming.swift
//  SwiftButtonsExample
//
//  Created by Eric Horacek on 3/14/15.
//  Copyright (c) 2015 Eric Horacek. All rights reserved.
//

import UIKit
import Motif

extension Themeable where ApplicantType == UIView {

    static func registerProperties() {
        ApplicantType.mtf_registerThemeProperty(
            ThemeProperties.backgroundColor.rawValue,
            requiringValueOf: UIColor.self,
            applierBlock: { (color, view, _) -> Bool in
                guard let view = view as? UIView else { return false }
                guard let color = color as? UIColor else { return false }

                view.backgroundColor = color

                return true
            })
        
        ApplicantType.mtf_registerThemeProperty(
            ThemeProperties.borderColor.rawValue,
            requiringValueOf: UIColor.self,
            applierBlock: { (color, view, _) -> Bool in
                guard let view = view as? UIView else { return false }
                guard let color = color as? UIColor else { return false }

                view.layer.borderColor = color.cgColor

                return true
            })
        
        ApplicantType.mtf_registerThemeProperty(
            ThemeProperties.cornerRadius.rawValue,
            requiringValueOf: NSNumber.self,
            applierBlock: { (cornerRadius, view, _) -> Bool in
                guard let view = view as? UIView else { return false }
                guard let cornerRadius = cornerRadius as? NSNumber else { return false }

                view.layer.cornerRadius = CGFloat(cornerRadius.floatValue)

                return true
            })
        
        ApplicantType.mtf_registerThemeProperty(
            ThemeProperties.borderWidth.rawValue,
            requiringValueOf: NSNumber.self,
            applierBlock: { (borderWidth, view, _) -> Bool in
                guard let view = view as? UIView else { return false }
                guard let borderWidth = borderWidth as? NSNumber else { return false }

                view.layer.borderWidth = CGFloat(borderWidth.floatValue)

                return true
            })
    }
    
}
